[ONNX] Add register empty custom op example

titaiwangms opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #106736


The empty custom opset is actually supported by onnx-script already. The demo uses a OnnxFunction to include one CustomOp.

The Onnxfunction is `foo_onnx`. It could be any custom function. NOTE: the eager mode is not supported by onnx-script in this case. as ORT doesn't recognize it.
<img width="242" alt="whole_graph" src="https://github.com/pytorch/pytorch/assets/18010845/575bde79-c295-4fb5-ae9f-61651566c55f">

Inside `foo_onnx`, there is one empty `CustomOp`.
<img width="149" alt="inside_func" src="https://github.com/pytorch/pytorch/assets/18010845/43495e7e-ce36-4a8d-8834-126a709e4593">

titaiwangms committed:
[ONNX] Add register empty custom op example

[ghstack-poisoned]

thiagocrepaldi commented on torch/onnx/_internal/exporter.py:
```diff
@@ -89,6 +89,44 @@ class OnnxRegistry:
+        >>> import torch.onnx
+        >>> from torch._custom_op import impl as custom_op
+        >>> import onnxscript
+        >>> @custom_op.custom_op("mylibrary::foo_op", "foo_op(Tensor x) -> Tensor")
+        >>> def foo_op(x):
```
We always encourage type hinting, but for custom ops API we have removed the type hint from the onnx function and added a string to `custom_op` with the expected typing. 

Is there a way to keep relying on type hinting? This is the most pythonic and intuitive way for users
What happens if the user do type hinting instead? Is there an error message to let them know where is the right place to declare type hint?
What happens if the user do type hinting and also use the type string on `custom_op`? Can we detect this scenario

titaiwangms replied:
> Might as well revert the workaround now that the example is skipped anyway due to `script`.
> 
Done

> On second thought, maybe we can demonstrate on aten op directly. The torch custom op took half the real estate here and is not our api.
> 
Use aten op to demo doesn't come in handy, because aten op is the target we will always try to record into torchlib, which means I will have to pop the example op out in the example, and that would be confused. I think we should stick to this example, and I put "Steps" to explain more what is goingo on between them. I think would help more. The very similar example is also written in test_fx_to_onnx.py to guard it.

> The `script` issue is problematic. It is expected for users to copy the example and run by themselves, If they run in interpreter directly it would fail.

I will see if I can understand the issue and fix. https://github.com/microsoft/onnxscript/issues/82

thiagocrepaldi replied:
The docstring is using type hinting, but both unit test files are still using the string argument at `custom_op`.

titaiwangms replied:
The test files should remain the way they are until `from __future__ import annotations` is supported in torch custom_op.

thiagocrepaldi replied:
ok, so the test must remain in the old way for now, but will the user reading the docstring be able to use the type hint style? Otherwise, we are publishing a docstring with non-functional code

titaiwangms replied:
Yes, they will. In fact, they should provide either type int or manual schema to create a custom op in torch. The best I can do is to revert back the test in `test_fx_pass.py` to type hint, and delete from __future__ import annotations, if we want to know if the custom op registration is changed in the future.

titaiwangms committed:
Update on "[ONNX] Add register empty custom op example"



The empty custom opset is actually supported by onnx-script already. The demo uses a OnnxFunction to include one CustomOp.

The Onnxfunction is `foo_onnx`. It could be any custom function. NOTE: the eager mode is not supported by onnx-script in this case. as ORT doesn't recognize it.
<img width="242" alt="whole_graph" src="https://github.com/pytorch/pytorch/assets/18010845/575bde79-c295-4fb5-ae9f-61651566c55f">

Inside `foo_onnx`, there is one empty `CustomOp`.
<img width="149" alt="inside_func" src="https://github.com/pytorch/pytorch/assets/18010845/43495e7e-ce36-4a8d-8834-126a709e4593">

 

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add register empty custom op example"



The empty custom opset is actually supported by onnx-script already. The demo uses a OnnxFunction to include one CustomOp.

The Onnxfunction is `foo_onnx`. It could be any custom function. NOTE: the eager mode is not supported by onnx-script in this case. as ORT doesn't recognize it.
<img width="242" alt="whole_graph" src="https://github.com/pytorch/pytorch/assets/18010845/575bde79-c295-4fb5-ae9f-61651566c55f">

Inside `foo_onnx`, there is one empty `CustomOp`.
<img width="149" alt="inside_func" src="https://github.com/pytorch/pytorch/assets/18010845/43495e7e-ce36-4a8d-8834-126a709e4593">

 

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add register empty custom op example"



The empty custom opset is actually supported by onnx-script already. The demo uses a OnnxFunction to include one CustomOp.

The Onnxfunction is `foo_onnx`. It could be any custom function. NOTE: the eager mode is not supported by onnx-script in this case. as ORT doesn't recognize it.
<img width="242" alt="whole_graph" src="https://github.com/pytorch/pytorch/assets/18010845/575bde79-c295-4fb5-ae9f-61651566c55f">

Inside `foo_onnx`, there is one empty `CustomOp`.
<img width="149" alt="inside_func" src="https://github.com/pytorch/pytorch/assets/18010845/43495e7e-ce36-4a8d-8834-126a709e4593">

 

[ghstack-poisoned]

titaiwangms commented:
Will put this in tutoral instead
